README.md
# [gulp](http://gulpjs.com)-responsive [![Build Status](https://travis-ci.org/mahnunchik/gulp-responsive.svg?branch=master)](https://travis-ci.org/mahnunchik/gulp-responsive)

[![Greenkeeper badge](https://badges.greenkeeper.io/mahnunchik/gulp-responsive.svg)](https://greenkeeper.io/)

> Generates images at different sizes

## Installation

`gulp-responsive` depends on [sharp](https://github.com/lovell/sharp). Sharp is one of the fastest Node.js modules for resizing JPEG, PNG, WebP and TIFF images.

If you are using Mac OS then before installing `gulp-responsive` you should install the [libvips](https://github.com/jcupitt/libvips) library. Further information and instructions can be found in the [sharp installation guide](http://sharp.dimens.io/en/stable/install/).

```sh
$ npm install --save-dev gulp-responsive
```

## Usage

```js
var gulp = require('gulp')
var responsive = require('gulp-responsive')

gulp.task('default', function () {
  return gulp
    .src('src/*.{png,jpg}')
    .pipe(
      responsive({
        'background-*.jpg': {
          width: 700,
          quality: 50
        },
        'cover.png': {
          width: '50%',
          // convert to jpeg format
          format: 'jpeg',
          rename: 'cover.jpg'
        },
        // produce multiple images from one source
        'logo.png': [
          {
            width: 200
          },
          {
            width: 200 * 2,
            rename: 'logo@2x.png'
          }
        ]
      })
    )
    .pipe(gulp.dest('dist'))
})
```

## [Examples](./examples)

- [Simple example](./examples/simple.md)
- [Multiple resolutions](./examples/multiple-resolutions.md)
- [Advanced example](./examples/advanced.md)

### Integration

- [HTML `srcset` attribute](./examples/srcset.md)
- [HTML `<picture>` element](./examples/picture.md)
- [CSS `image-set` method](./examples/image-set.md)

### All together :fireworks:

- [`gulp-responsive` config generation example](./examples/gulp-responsive-config.md)

## API

### responsive([config](#config), [options](#options))

#### config

Configuration can be provided in one of the following formats:

##### 1. Array of unique configurations

```js
;[
  {
    name: 'logo.png',
    width: 200,
    height: 100
  },
  {
    name: 'banner.png',
    width: 500
  }
]
```

##### 2. Object of unique configurations. Keys are filenames.

```js
{
  'logo.png': {
    width: 300,
    height: 200,
    rename: 'logo@2x.png'
  },
  'background-*.png': {
    width: 1400,
    withoutEnlargement: true
  }
}
```

##### 3. Object of arrays of unique configurations. Keys are filenames.

```js
{
  'logo.png': [{
      width: 200,
      rename: 'logo@1x.png'
    },{
      width: 400,
      rename: 'logo@2x.png'
    }],
  'background-*': [{
    height: 400
  }]
}
```

#### Configuration unit

Configuration unit is an object:

- **name**: _String_ — filename glob pattern.
- **width**: _Number_ or _String_ — width in pixels or percentage of the original, not set by default.
- **height**: _Number_ or _String_ — height in pixels or percentage of the original, not set by default.
- [**withoutEnlargement**](http://sharp.dimens.io/en/stable/api-resize/#withoutenlargement): _Boolean_ — do not enlarge the output image, default `true`.
- **skipOnEnlargement**: _Boolean_ — do not write an output image at all if the original image is smaller than the configured width or height, default `false`.
- [**min**](http://sharp.dimens.io/en/stable/api-resize/#min): _Boolean_ — preserving aspect ratio, resize the image to be as small as possible while ensuring its dimensions are greater than or equal to the `width` and `height` specified.
- [**max**](http://sharp.dimens.io/en/stable/api-resize/#max): _Boolean_ — resize to the max width or height the preserving aspect ratio (both `width` and `height` have to be defined), default `false`.
- [**quality**](http://sharp.dimens.io/en/stable/api/#qualityquality): _Number_ — output quality for JPEG, WebP and TIFF, default `80`.
- [**progressive**](http://sharp.dimens.io/en/stable/api/#progressive): _Boolean_ — progressive (interlace) scan for JPEG and PNG output, default `false`.
- [**withMetadata**](http://sharp.dimens.io/en/stable/api-output/#withmetadata): _Boolean_ — include image metadata, default `false`.
- [**compressionLevel**](http://sharp.dimens.io/en/stable/api/#compressionlevelcompressionlevel): _Number_ — zlib compression level for PNG, default `6`.
- [**rename**](#renaming): _String_, _Object_ or _Function_ — renaming options, file will not be renamed by default. When `extname` is specified, output format is parsed from extension. You can override this autodetection with `format` option.
- [**format**](http://sharp.dimens.io/en/stable/api-output/#toformat): _String_ — output format `jpeg`, `png`, `webp` or `raw`, default is `null`.
- [**crop**](http://sharp.dimens.io/en/stable/api-resize/#crop): Crop the resized image to the exact size specified, default is `false`.
- [**embed**](http://sharp.dimens.io/en/stable/api-resize/#embed): Preserving aspect ratio, resize the image to the maximum `width` or `height` specified then `embed` on a `background` of the exact `width` and `height` specified, default is `false`.
- [**ignoreAspectRatio**](http://sharp.dimens.io/en/stable/api-resize/#ignoreaspectratio): _Boolean_ — Ignoring the aspect ratio of the input, stretch the image to the exact `width` and/or `height` provided via `resize`, default is `false`.
- [**kernel**](http://sharp.dimens.io/en/stable/api/#resizewidth-height-options): _String_ — The kernel to use for image **reduction**, defaulting to `lanczos3`.
- [**background**](http://sharp.dimens.io/en/stable/api-colour/#background): [_Color_](https://www.npmjs.com/package/color) — Set the background for the embed and flatten operations, '#default is `fff`'.
- [**flatten**](http://sharp.dimens.io/en/stable/api-operation/#flatten): _Boolean_ — Merge alpha transparency channel, if any, with `background`, default is `false`.
- [**negate**](http://sharp.dimens.io/en/stable/api-operation/#negate): _Boolean_ — Produces the "negative" of the image, default is `false`.
- [**rotate**](http://sharp.dimens.io/en/stable/api-operation/#rotate): _Boolean_ — Rotate the output image by either an explicit angle or auto-orient based on the EXIF `Orientation` tag, default is `false`.
- [**flip**](http://sharp.dimens.io/en/stable/api-operation/#flip): _Boolean_ — Flip the image about the vertical Y axis. This always occurs after rotation, if any. The use of `flip` implies the removal of the EXIF `Orientation` tag, if any. Default is `false`.
- [**flop**](http://sharp.dimens.io/en/stable/api-operation/#flop): _Boolean_ — Flop the image about the horizontal X axis. This always occurs after rotation, if any. The use of `flop` implies the removal of the EXIF `Orientation` tag, if any. Default is `false`.
- [**blur**](http://sharp.dimens.io/en/stable/api-operation/#blur): _Boolean_ — When used without parameters, performs a fast, mild blur of the output image. This typically reduces performance by 10%. Default is `false`.
- [**sharpen**](http://sharp.dimens.io/en/stable/api-operation/#sharpen): _Boolean_ — When used without parameters, performs a fast, mild sharpen of the output image. This typically reduces performance by 10%. Default is `false`.
- [**threshold**](http://sharp.dimens.io/en/stable/api-operation/#threshold): _Number_ or _Boolean_ — Converts all pixels in the image to greyscale white or black, default is `false`.
- [**gamma**](http://sharp.dimens.io/en/stable/api-operation/#gamma): _Boolean_ — Apply a gamma correction by reducing the encoding (darken) pre-resize at a factor of `1/gamma` then increasing the encoding (brighten) post-resize at a factor of `gamma`. Default is `false`.
- [**grayscale**](http://sharp.dimens.io/en/stable/api-colour/#greyscale): _Boolean_ — Convert to 8-bit greyscale; 256 shades of grey, default is `false`.
- [**normalize**](http://sharp.dimens.io/en/stable/api-operation/#normalise): _Boolean_ — Enhance output image contrast by stretching its luminance to cover the full dynamic range. This typically reduces performance by 30%. Default is `false`.
- [**trim**](http://sharp.dimens.io/en/stable/api-operation/#trim): _Boolean_ or _Number_ — Trim "boring" pixels from all edges that contain values within a percentage similarity of the top-left pixel. Default is `false`.
- [**tile**](http://sharp.dimens.io/en/stable/api-output/#tile): _Boolean_ or _Object_ — The size and overlap, in pixels, of square Deep Zoom image pyramid tiles, default is `false`.
- [**withoutChromaSubsampling**](http://sharp.dimens.io/en/stable/api/#withoutchromasubsampling): _Boolean_ — Disable the use of [chroma subsampling](http://en.wikipedia.org/wiki/Chroma_subsampling) with JPEG output (4:4:4), default is `false`.

Detailed description of each option can be found in the [sharp API documentation](http://sharp.dimens.io/en/stable/api/).

##### Renaming

Renaming is implemented by the [rename](https://www.npmjs.com/package/rename) module. Options correspond with options of [gulp-rename](https://www.npmjs.com/package/gulp-rename).

#### options

##### errorOnUnusedConfig

Type: `Boolean`  
Default: `true`

Emit the error when configuration is not used.

##### errorOnUnusedImage

Type: `Boolean`  
Default: `true`

Emit the error when image is not used.

##### passThroughUnused

Type: `Boolean`  
Default: `false`

Keep unmatched images in the stream.
To use this option `errorOnUnusedImage` should be `false`.

##### errorOnEnlargement

Type: `Boolean`  
Default: `true`

Emit the error when image is enlarged.

##### stats

Type: `Boolean`  
Default: `true`

Show statistics after the run — how many images were created, how many were matched and how many were in the run in total.

##### silent

Type: `Boolean`  
Default: `false`

Silence messages and stats if 0 images were created. If you wish to supress all messages and stats, set the `options.stats` to `false` as well.

> You can specify **global default value** for any of the [configuration options](#configuration-unit).

```js
gulp.task('default', function () {
  return gulp
    .src('src/*.png')
    .pipe(
      responsive(config, {
        // global quality for all images
        quality: 50,
        errorOnUnusedImage: false
      })
    )
    .pipe(gulp.dest('dist'))
})
```

## License

MIT © [Evgeny Vlasenko](https://github.com/mahnunchik)


History.md

# 3.0.0 / 2019-10-12

- Suport for sharp `0.23.1` [#129](https://github.com/mahnunchik/gulp-responsive/pull/129)
- Remove support for Node.js 6 [#129](https://github.com/mahnunchik/gulp-responsive/pull/129)
- Support for Node.js 8,10,12 [#129](https://github.com/mahnunchik/gulp-responsive/pull/129)
- Refactor codebase [#129](https://github.com/mahnunchik/gulp-responsive/pull/129)
- Avoid renaming extname [#116](https://github.com/mahnunchik/gulp-responsive/pull/116)
- Fix value for chroma subsampling setting [#116](https://github.com/mahnunchik/gulp-responsive/pull/116)

# 2.6.0 / 2016-09-15

- add: `kernel` option
- rename: `interpolation` to `interpolator` option
- deprecate: `interpolation` option

  # 2.5.0 / 2016-06-01

- deps: sharp@0.16.0 http://sharp.dimens.io/en/stable/changelog/#v016-pencil
- deps: async@2.0.1
- deps: mocha@3.0.2

  # 2.4.0 / 2016-06-01

- deps: sharp@0.15.0 http://sharp.dimens.io/en/stable/changelog/#v015-outfit
- breaking: Existing sharpen API to accept sigma

  # 2.3.0 / 2016-04-15

- deps: sharp@0.14.0 http://sharp.dimens.io/en/stable/changelog/#v014-needle

  # 2.2.0 / 2016-03-18

- add: silent option
- add: stats option

  # 2.1.0 / 2016-02-18

- deps: sharp@0.13.0
- deps: lodash@4.5

  # 2.0.0 / 2016-01-09

- Version 2.0 https://github.com/mahnunchik/gulp-responsive/issues/23
- add: all sharp options
- add: better usage examples
- add: ability to transform format of output image (`format` option)
- remove: depricated options `strictMatchConfig` and `strictMatchImages`
- add: visibility to image manipulation process

  # 1.2.1 / 2015-08-18

- add: `skipOnEnlargement` option

  # 1.2.0 / 2015-08-15

- deps: sharp@0.11 through2@2

  # 1.1.0 / 2015-05-24

- add: `passThroughUnused` option to pass through unmatched files
- rename: options `strictMatchConfig` to `errorOnUnusedConfig` and `strictMatchImages` to `errorOnUnusedImage`
- deps: sharp@^0.10.0, rename@^1.0.3, async@^1.0.0

  # 1.0.7 / 2015-02-16

- add: `max` option, see [sharp#max](https://github.com/lovell/sharp#max)

  # 1.0.6 / 2015-02-08

- add: node 0.12 and latest iojs to travis

  # 1.0.5 / 2015-01-28

- update: sharp 0.9.0
- update: minimatch and lodash

  # 1.0.4 / 2014-11-23

- add: ability to use size as a percentage

  # 1.0.3 / 2014-11-19

- add: global config feature

  # 1.0.2 / 2014-11-11

- add: renaming feature by `rename` module https://www.npmjs.org/package/rename

# 1.0.1 / 2014-11-03

- add: config options: quality, progressiv, withMetadata, compressionLevel
- fix: merge config behavior

# 1.0.0 / 2014-10-29

- basic implementation


test.js
'use strict';
var assert = require('assert');
var gutil = require('gulp-util');
var responsive = require('./');

it('should ', function (cb) {
	var stream = responsive();

	stream.on('data', function (file) {
		assert.equal(file.relative, 'file.ext');
		assert.equal(file.contents.toString(), 'unicorns');
	});

	stream.on('end', cb);

	stream.write(new gutil.File({
		base: __dirname,
		path: __dirname + '/file.ext',
		contents: new Buffer('unicorns')
	}));

	stream.end();
});


test/size.js
/* global describe, it */

'use strict'

const assert = require('assert')
const size = require('../lib/size')
const _ = require('lodash')

describe('gulp-responsive', function () {
  describe('size parser', function () {
    it('should leave number unchanged', function () {
      const test = size(100)
      assert.strictEqual(test, 100)
    })

    it('should leave null', function () {
      const test = size()
      assert(test === null)
    })

    it('should parse string as a number', function () {
      const test = size('100')
      assert.strictEqual(test, 100)
      assert(_.isNumber(test))
    })

    it('should parse string in pixels as a number', function () {
      const test = size('100px')
      assert.strictEqual(test, 100)
      assert(_.isNumber(test))
    })

    it('should recognize percentages and calculate size', function () {
      const test = size('10%', 900)
      assert.strictEqual(test, 90)
      assert(_.isNumber(test))
    })

    it('should recognize percentages and calculate size (more then 100%)', function () {
      const test = size('200%', 900)
      assert.strictEqual(test, 1800)
      assert(_.isNumber(test))
    })

    it('should throw an error on wrong input', function () {
      assert.throws(function () {
        size('wrong')
      }, /Wrong size/)
    })

    it('should throw an error on wrong percentage input', function () {
      assert.throws(function () {
        size('wrong%')
      }, /Wrong percentage size/)
    })
  })
})


test/test.js
/* global describe, it */

'use strict'

const assert = require('assert')
const path = require('path')

const responsive = require('../')

const helpers = require('./helpers')
const makeFile = helpers.makeFile
const assertFile = helpers.assertFile

describe('gulp-responsive', function () {
  it('should not do anything without images and configs', function (cb) {
    const stream = responsive()

    stream.on('end', cb)
    stream.on('data', function () {
      throw new Error('data should not be provided')
    })

    stream.end()
  })

  it('should provide one image when exactly one image and one config are provided', function (cb) {
    const config = [
      {
        name: 'gulp.png'
      }
    ]
    const stream = responsive(config)

    let counter = 0

    stream.on('data', function (file) {
      counter++
      assertFile(file)
      if (counter > 1) {
        throw new Error('more than one file is provided')
      }
    })

    stream.on('end', function () {
      assert.strictEqual(counter, 1)
      cb()
    })

    stream.write(makeFile('gulp.png'))
    stream.end()
  })

  it('should support source file in SVG format', function (cb) {
    const config = [
      {
        name: 'gulp.svg',
        format: 'png'
      }
    ]
    const stream = responsive(config)

    let counter = 0

    stream.on('data', function (file) {
      counter++
      assertFile(file)
      if (counter > 1) {
        throw new Error('more than one file is provided')
      }
    })

    stream.on('end', function () {
      assert.strictEqual(counter, 1)
      cb()
    })

    stream.write(makeFile('gulp.svg'))
    stream.end()
  })

  it('should provide two image when one image and exactly two configs are provided', function (cb) {
    const config = [
      {
        name: 'gulp.png'
      },
      {
        name: 'gulp.png'
      }
    ]
    const stream = responsive(config)

    let counter = 0

    stream.on('data', function (file) {
      counter++
      assertFile(file)
      if (counter > 2) {
        throw new Error('more than two files are provided')
      }
    })

    stream.on('end', function () {
      assert.strictEqual(counter, 2)
      cb()
    })

    stream.write(makeFile('gulp.png'))
    stream.end()
  })

  it('should provide two image when one image match two configs', function (cb) {
    const config = [
      {
        name: 'gulp.png'
      },
      {
        name: '*.png'
      }
    ]
    const stream = responsive(config)

    let counter = 0

    stream.on('data', function (file) {
      counter++
      assertFile(file)
      if (counter > 2) {
        throw new Error('more than two files are provided')
      }
    })

    stream.on('end', function () {
      assert.strictEqual(counter, 2)
      cb()
    })

    stream.write(makeFile('gulp.png'))
    stream.end()
  })

  describe('rename image', function () {
    it('should provide renamed image when rename is string', function (cb) {
      const config = [
        {
          name: 'gulp.png',
          rename: 'test.png'
        }
      ]
      const stream = responsive(config)

      stream.on('data', function (file) {
        assertFile(file)
        assert.strictEqual(
          file.path,
          path.join(__dirname, '/fixtures/', 'test.png')
        )
      })

      stream.on('end', function () {
        cb()
      })

      stream.write(makeFile('gulp.png'))
      stream.end()
    })

    it('should provide renamed image when rename is object', function (cb) {
      const config = [
        {
          name: 'gulp.png',
          rename: {
            suffix: '-renamed'
          }
        }
      ]
      const stream = responsive(config)

      stream.on('data', function (file) {
        assertFile(file)
        assert.strictEqual(
          file.path,
          path.join(__dirname, '/fixtures/', 'gulp-renamed.png')
        )
      })

      stream.on('end', function () {
        cb()
      })

      stream.write(makeFile('gulp.png'))
      stream.end()
    })

    it('should provide renamed image when rename is function', function (cb) {
      const config = [
        {
          name: 'gulp.png',
          rename: function (path) {
            path.basename += '-renamed-by-function'
            return path
          }
        }
      ]
      const stream = responsive(config)

      stream.on('data', function (file) {
        assertFile(file)
        assert.strictEqual(
          file.path,
          path.join(__dirname, '/fixtures/', 'gulp-renamed-by-function.png')
        )
      })

      stream.on('end', function () {
        cb()
      })

      stream.write(makeFile('gulp.png'))
      stream.end()
    })
  })

  describe('unmatched/unused images', function () {
    it('should not pass through unmatched file by default when `errorOnUnusedImage` is false', function (cb) {
      const stream = responsive(
        {},
        {
          errorOnUnusedImage: false
        }
      )

      let counter = 0

      stream.on('data', function () {
        counter++
      })

      stream.on('end', function () {
        assert.strictEqual(counter, 0)
        cb()
      })

      stream.write(makeFile('gulp.png'))
      stream.end()
    })

    it('should pass through unmatched file when `passThroughUnused` is true and `errorOnUnusedImage` is false', function (cb) {
      const expectedFile = makeFile('gulp.png')

      const stream = responsive(
        {},
        {
          errorOnUnusedImage: false,
          passThroughUnused: true
        }
      )

      let counter = 0

      stream.on('data', function (file) {
        counter++
        if (counter > 1) {
          throw new Error('more than two files are provided')
        }
        assertFile(file)
        assert.deepStrictEqual(file, expectedFile)
      })

      stream.on('end', function () {
        assert.strictEqual(counter, 1)
        cb()
      })

      stream.write(expectedFile)
      stream.end()
    })

    it('should skip enlarged image when `skipOnEnlargement` is true', function (cb) {
      const config = [
        {
          name: 'gulp.png',
          width: 10000
        }
      ]

      const stream = responsive(config, {
        errorOnEnlargement: false,
        skipOnEnlargement: true
      })

      stream.on('data', function () {
        throw new Error('enlarged image not been skipped')
      })

      stream.on('end', function () {
        cb()
      })

      stream.write(makeFile('gulp.png'))
      stream.end()
    })
  })
})


lib/index.js
'use strict'

const chalk = require('chalk')
const flog = require('fancy-log')
const through2 = require('through2')
const minimatch = require('minimatch')
const PluginError = require('plugin-error')
const plur = require('plur')

const sharpVinyl = require('./sharp')
const prepareConfig = require('./config')

const PLUGIN_NAME = require('../package.json').name

function gulpResponsive (config, options) {
  const statistics = {
    total: 0,
    matched: 0,
    created: 0,
    unmatched: 0,
    unmatchedBlocked: 0,
    unmatchedPassed: 0
  }

  options = {
    errorOnUnusedConfig: true,
    errorOnUnusedImage: true,
    errorOnEnlargement: true,
    passThroughUnused: false,
    silent: false,
    stats: true,
    ...options
  }

  config = prepareConfig(config || [], options)

  return through2.obj(
    async function through2Handler (file, enc, done) {
      const that = this

      if (file.isNull()) {
        this.push(file)
        return done()
      }

      if (file.isStream()) {
        return done(new PluginError(PLUGIN_NAME, 'Streaming not supported'))
      }

      statistics.total++
      const matched = config.filter(conf => minimatch(file.relative, conf.name))

      if (matched.length === 0) {
        statistics.unmatched++
        const message =
          'File `' + file.relative + '`: Image does not match any config'
        if (options.errorOnUnusedImage) {
          return done(new PluginError(PLUGIN_NAME, message))
        } else if (options.passThroughUnused) {
          this.push(file)
          statistics.unmatchedPassed++
          if (!options.silent) {
            flog(
              PLUGIN_NAME + ': (pass through without changes)',
              chalk.magenta(message)
            )
          }
          return done()
        }
        statistics.unmatchedBlocked++
        if (!options.silent) {
          flog(PLUGIN_NAME + ': (skip for processing)', chalk.magenta(message))
        }
        return done()
      }

      statistics.matched++

      const allPromises = []

      for (const conf of matched) {
        // config item matched (can be matched multiple times)
        conf.matched = true
        allPromises.push(sharpVinyl(file, conf, options))
      }

      try {
        const result = await Promise.all(allPromises)

        result.forEach(newFiles => {
          if (newFiles) {
            newFiles.forEach(f => that.push(f))
            statistics.created += newFiles.length
          }
        })
        done()
      } catch (error) {
        done(new PluginError(PLUGIN_NAME, error, { showStack: true }))
      }
    },
    function (cb) {
      const notMatched = config.filter((conf) => !conf.matched)

      if (options.stats && !(options.silent && statistics.created === 0)) {
        const msg =
          'Created ' +
          statistics.created +
          ' ' +
          plur('image', statistics.created) +
          chalk.dim.white(
            ' (matched ' +
              statistics.matched +
              ' of ' +
              statistics.total +
              ' ' +
              plur('image', statistics.total) +
              ')'
          )

        flog(PLUGIN_NAME + ':', chalk.green(msg))
      }

      if (
        notMatched.length > 0 &&
        (!options.silent || options.errorOnUnusedConfig)
      ) {
        let message = 'Available images do not match the following config:'
        notMatched.forEach(function (conf) {
          message += '\n  - `' + conf.name + '`'
        })
        if (options.errorOnUnusedConfig) {
          return cb(new PluginError(PLUGIN_NAME, message))
        } else {
          flog(PLUGIN_NAME + ':', chalk.magenta(message))
        }
      }
      cb()
    }
  )
}

module.exports = gulpResponsive


lib/sharp.js
'use strict'

const sharp = require('sharp')
const chalk = require('chalk')
const flog = require('fancy-log')
const path = require('path')
const PluginError = require('plugin-error')
const rename = require('rename')
const Vinyl = require('vinyl')
const size = require('./size')
const format = require('./format')

const PLUGIN_NAME = require('../package.json').name

module.exports = async function sharpVinyl (file, config, options) {
  const errPrefix = 'File `' + file.relative + '`: '
  const image = sharp(file.contents)

  try {
    const metadata = await image.metadata()
    let extract, toFormat

    let filePath = file.path
    if (config.rename) {
      filePath = path.join(file.base, rename(file.relative, config.rename))
    }

    const width = size(config.width, metadata.width)
    const height = size(config.height, metadata.height)

    if (width || height) {
      if (
        config.withoutEnlargement &&
        (width > metadata.width || height > metadata.height)
      ) {
        let message = errPrefix + 'Image enlargement is detected'
        if (width) {
          message +=
            '\n  real width: ' +
            metadata.width +
            'px, required width: ' +
            width +
            'px'
        }
        if (height) {
          message +=
            '\n  real height: ' +
            metadata.height +
            'px, required height: ' +
            height +
            'px'
        }
        if (options.errorOnEnlargement) {
          throw new PluginError(PLUGIN_NAME, message)
        } else if (config.skipOnEnlargement) {
          if (!options.silent) {
            flog(PLUGIN_NAME + ': (skip for processing)', chalk.red(message))
          }
          // passing a null file to the callback stops a new image being added to the pipeline for this config
          return null
        }
        if (!options.silent) {
          flog(PLUGIN_NAME + ': (skip for enlargement)', chalk.yellow(message))
        }
      }
    }

    if (config.extractBeforeResize) {
      extract = config.extractBeforeResize
      image.extract(extract)
    }

    image.resize(width, height, {
      background: config.background,
      kernel: config.kernel,
      withoutEnlargement: config.withoutEnlargement
    })

    if (config.extractAfterResize) {
      extract = config.extractAfterResize
      image.extract(extract)
    }

    if (config.crop !== false) {
      if (config.crop === 'entropy') {
        image.crop(sharp.strategy.entropy)
      } else if (config.crop === 'attention') {
        image.crop(sharp.strategy.attention)
      } else {
        image.crop(config.crop)
      }
    }

    if (config.embed) {
      image.embed()
    }

    if (config.max) {
      image.max()
    }

    if (config.flatten) {
      image.flatten({
        background: config.background
      })
    }

    image.negate(config.negate)

    if (config.trim) {
      image.trim(config.trim)
    }

    if (config.min) {
      image.min()
    }

    if (config.ignoreAspectRatio) {
      image.ignoreAspectRatio()
    }

    image.flatten(config.flatten)
    image.negate(config.negate)

    if (config.rotate !== false) {
      if (typeof config.rotate === 'boolean') {
        image.rotate()
      } else {
        image.rotate(config.rotate)
      }
    }

    image.flip(config.flip)
    image.flop(config.flop)
    image.blur(config.blur)

    if (typeof config.sharpen === 'boolean') {
      image.sharpen(config.sharpen)
    } else {
      image.sharpen(
        config.sharpen.sigma,
        config.sharpen.flat,
        config.sharpen.jagged
      )
    }

    image.threshold(config.threshold)

    if (config.gamma !== false) {
      if (typeof config.gamma === 'boolean') {
        image.gamma()
      } else {
        image.gamma(config.gamma)
      }
    }

    image.grayscale(config.grayscale)
    image.normalize(config.normalize)
    image.withMetadata(config.withMetadata)
    image.tile(config.tile)

    if (config.withoutChromaSubsampling) {
      config.chromaSubsampling = '4:4:4'
    }

    if (config.format) {
      toFormat = config.format
    } else {
      toFormat = format(filePath)
    }

    if (!Array.isArray(toFormat)) {
      toFormat = [toFormat]
    }

    const allOps = toFormat.map(async format => {
      const clone = convertImage(image, format, config)
      const buffer = await clone.toBuffer()

      const newFile = new Vinyl({
        cwd: file.cwd,
        base: file.base,
        path: filePath,
        contents: buffer
      })

      newFile.extname = '.' + format

      if (!options.silent) {
        flog(
          PLUGIN_NAME + ':',
          chalk.green(file.relative + ' -> ' + newFile.relative)
        )
      }

      return newFile
    })

    return Promise.all(allOps)
  } catch (error) {
    throw new PluginError(PLUGIN_NAME, errPrefix + error.message, { showStack: true })
  }
}

function convertImage (image, format, config) {
  const clone = image.clone()

  switch (format) {
    case 'jpeg':
    case 'jpg':
    case 'jpe':
      return clone.jpeg(config)
    case 'png':
    case 'webp':
    case 'tiff':
    default:
      return clone.toFormat(format, config)
  }
}


lib/config.js
/* jshint -W083 */
'use strict'

/*
Configuration unit is an object:

* name: String - filename glob pattern
* width: Number - not set by default
* height: Number - not set by default
* withoutEnlargement: Boolean - default true
* skipOnEnlargement: Boolean - default false
* rename: String - new file name, file will not be renamed by dafault

Configuration can be provided in one of the following formats:

1. Array of unique configurations

[{
  name: 'logo.png',
  width: 200,
  height: 100
},{
  name: 'banner.png',
  width: 500
}]

2. Object of unique configurations. Keys are names of files.

{
  'logo.png': {
    width: 300,
    height: 200,
    raname: 'logo@2x.png'
  },
  'background-*.png': {
    width: 1400,
    withoutEnlargement: true
  }
}

3. Object of array of unique configurations. Keys are names of files.

{
  'logo.png': [{
      width: 200,
      raname: 'logo@1x.png'
    },{
      width: 400,
      rename: 'logo@2x.png'
    }],
  'background-*': [{
    height: 400
  }]
}

*/

const defaultConfig = {
  crop: false,
  embed: false,
  min: false,
  max: false,
  withoutEnlargement: true,
  skipOnEnlargement: false,
  ignoreAspectRatio: false,
  kernel: 'lanczos3',
  extractBeforeResize: false,
  extractAfterResize: false,
  background: '#fff',
  flatten: false,
  negate: false,
  rotate: false,
  flip: false,
  flop: false,
  blur: false,
  sharpen: false,
  threshold: false,
  gamma: false,
  grayscale: false,
  normalize: false,
  quality: 80,
  progressive: false,
  withMetadata: false,
  tile: false,
  withoutChromaSubsampling: false,
  compressionLevel: 6,
  format: null,
  trim: false
}

function prepareConfig (config, globalConfig) {
  const preparedConfig = []

  if (!Array.isArray(config)) {
    for (const name in config) {
      if (Object.prototype.hasOwnProperty.call(config, name)) {
        if (Array.isArray(config[name])) {
          config[name].forEach(function (conf) {
            preparedConfig.push({ ...defaultConfig, ...globalConfig, ...conf, name })
          })
        } else {
          preparedConfig.push({ ...defaultConfig, ...globalConfig, ...config[name], name })
        }
      }
    }
  } else {
    config.forEach(conf => {
      preparedConfig.push({ ...defaultConfig, ...globalConfig, ...conf })
    })
  }
  return preparedConfig
}

module.exports = prepareConfig
